### ISSUE TITLE
Help reproducing AP from paper

### ISSUE THREAD

**Issue by sammyKhan** [Noble-Lab/casanovo#615]
Hi there,

We’ve been working on integrating a new module into Casanovo. We’d like to publish our results, but are having a hard time reproducing the exact baseline results from [S2](https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-024-49731-x/MediaObjects/41467_2024_49731_MOESM1_ESM.pdf), which we are building off of.

From our perspective it would be fine to simply use the numbers we’re getting, since it’s the delta that’s we're most interested in anyways. We would then simply acknowledge that our baseline doesn’t exactly match the reported values (could be a number of reasons including beam search decoding, min_peaks etc which are not reported in the paper).

That said, our preference would be to have a baseline that matches, so any pointers for where the discrepancy might come from would be greatly appreciated! We’re calculating the AP as outlined  in the [FAQ](https://casanovo.readthedocs.io/en/latest/faq.html#how-can-i-generate-a-precisioncoverage-curve). I'm using the [revised benchmark](https://zenodo.org/records/10939843) and all defaults from GitHub (most recent release, v5.0.0 weights).

The AP I'm calculating for human is 0.915 , and seeing similar discrepancies for the other species (except mouse).
Here are the mztab and logfile:

[9s_human_dnps_FRESH_INSTALL.log](https://github.com/user-attachments/files/26456225/9s_human_dnps_FRESH_INSTALL.log)

[9s_human_dnps_FRESH_INSTALL.mztab.gz](https://github.com/user-attachments/files/26456247/9s_human_dnps_FRESH_INSTALL.mztab.gz)

MGF input (can't upload to ticket due to size limit): https://drive.google.com/file/d/1hyXY5m8x9zFRC2TaX_hgbcGGOhhwDRyk/view?usp=drive_link

**Comment by wsnoble** [comment]
Thanks for your interest in Casanovo and for the detailed description of what you did to reproduce our supplementary figure.  I've reached out to Melih to see if we can locate that actual code that generated that figure.  

One clear difference is that you are using v5.0.0 weights, which were released after the Nat Comms paper was published.  You might need to use v4.2.0 weights to reproduce those results.

I looked at the log file you sent and didn't see anything unusual.  I believe that S2 should have been generated using the default parameters, which seems to be what you are doing as well. 

One issue could be how long the model trained. Do you have a plot of the train and validation loss?  Did 30 epochs take you beyond the first minimum in the cosine scheduler?

**Comment by sammyKhan** [comment]
Thanks for the quick response, I appreciate your help!

I am actually only doing inference using the released weights, not doing my own training.
I could rerun with the v4.2.0 weights, but that would indicate a major performance regression, so I rather suspect it's a data issue instead.

I think I found one piece of the puzzle: Looking at the [Source Data](https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-024-49731-x/MediaObjects/41467_2024_49731_MOESM4_ESM.zip) attached to the paper, we find the .xlsx that was used for the calculations. For human, the first coverage value is 0.00002914602157, indicating that the number of PSMs evaluated were 1/0.00002914602157 = 34310. When I'm running it, I'm getting >44k PSMs in the human part of the nine-species benchmark, as also reported in [table 1](https://www.nature.com/articles/s41597-024-04068-4/tables/1) of the dataset paper.

**Comment by bittremieux** [comment]
Is the difference that you're observing an AP of 0.915 vs 0.94 reported in that figure? In that case I wouldn't call it a "major performance regression", and it could indeed just be the difference between the v4.2 and v5.0 weights.

**Comment by wsnoble** [comment]
Sorry, I misremembered the version number.  Melih said Casanova 4.0.1 was used (as noted in the paper) and we were using n_beams=5.

**Comment by sammyKhan** [comment]
Thanks for your continued support! I reran using Casanovo 4.0.1 and n_beams=5:

```
$ conda create --name khsam_casanovov401
$ conda activate khsam_casanovov401
$ conda install pip
$ pip install casanovo==4.0.1
$ casanovo configure
$ sed -i 's/weights_only=weights_only/weights_only=False/g' /opt/modules/i12g/anaconda/envs/khsam_casanovov401/lib/python3.11/site-packages/lightning/pytorch/core/saving.py
$ sed -i 's/n_beams: 1/n_beams: 5/g' ./casanovo.yaml
$ casanovo sequence -m /s/project/denovo-prosit/CASANOVO/casanovo_4.1_model_weights/casanovo_massivekb_v4_0_0.ckpt -c ./casanovo.yaml -o casanovo_v401.mztab /s/project/denovo-prosit/data/benchmark_data/nine_species_benchmark_noble/nine-species-main/H.-sapiens/*
```

This is the resulting file:

[casanovo_v401.log](https://github.com/user-attachments/files/26490696/casanovo_v401.log)
[casanovo_v401.mztab.gz](https://github.com/user-attachments/files/26490695/casanovo_v401.mztab.gz)

For which I calculate AP=0.901

@bittremieux with AP=0.901, the final precision is 0.64 compared to 0.74 at AP=0.94, so it's a visually striking difference in the PC curve. I'm focusing on human in this ticket to make it easier to debug, but am encountering similar results for the other species.

Could you double-check the observation in my last post, that there appear to have been (>20% for human) PSMs excluded from the S2 curves? From my understanding it wouldn't have been `the revised nine-species benchmark with new modifications eliminated`, since that subset should have AP=0.98, but perhaps a different subset along those lines. What I could imagine, for example, is that some of the other models being benchmarked produced N/A values for some inputs, and that these rows were then mistakenly filtered out of the final data

**Comment by wsnoble** [comment]
Sorry for the slow response -- this took a while to track down.  I found the data on an external drive and contacted the first author.  He, together with Claude sleuthing through his old directories, confirmed what I suspected: we removed from the final curve any spectrum for which at least one of the competing methods failed to make a prediction.  This included all spectra containing PTMs, since some methods can't handle those.

Here is what Claude reported:

The filtering occurs in eval_output_for_all_models.ipynb, cell ee18e6f3:
                                                                                                                                                                                            
  `output_df = pd.read_csv(output_file).dropna(subset=['dn_output_seq']).reset_index(drop=True)  `                                                                                            
                                                                                                                                                                                            
  This drops every spectrum for which DeepNovo failed to produce a prediction. That's the source of the discrepancy.                                                                        
                  
  The full pipeline is:

  1. align_output_for_all_models.ipynb → merges all 4 models' outputs into all_model_outputs/{species}_all_models.csv. Notably, a dropna on dn_output_seq was considered but is commented   out (cell 22ba67d8).
  2. eval_output_for_all_models.ipynb → reads those files and actively applies .dropna(subset=['dn_output_seq']), silently discarding any spectrum DeepNovo skipped. It then writes the evaluated results to all_model_results/{subfolder}/{species}_all_models.csv.                                                                                                              
  3. plot_pc_curves.ipynb → reads from all_model_results/ and generates the final CSVs and plots. By this point the filtering has already happened.
                                                                                                                                                                                            
  There's also a second filter: when subfolder = "casanovo_only_ptm_excl" (which is what was used for the paper plots), this additional line runs:                                          
  `output_df = output_df[~output_df.apply(lambda row: ptm_in_pep(row.true_seq), axis=1)]`                                                                                                     
  This removes spectra where the ground-truth peptide contains PTMs like +42.011, +43.006, -17.027, N+0.984, or Q+0.984. The combination of both filters — dropping spectra with no DeepNovo   prediction AND dropping PTM-containing ground truths — is what reduces >44k PSMs to ~34,310.    

A zip file containing the CSVs and .py scripts to generate the figure is [here](https://noble.gs.washington.edu/~wnoble/temp/revised_benchmark.zip).

[align_output_for_all_models.ipynb](https://github.com/user-attachments/files/27028923/align_output_for_all_models.ipynb)
[eval_output_for_all_models.ipynb](https://github.com/user-attachments/files/27028922/eval_output_for_all_models.ipynb)
[plot_pc_curves.ipynb](https://github.com/user-attachments/files/27028921/plot_pc_curves.ipynb)
